core/server/worker/src/main/java/alluxio/worker/page/PagedUfsBlockReader.java

```diff
@@ -0,0 +1,303 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.worker.page;
+
+import alluxio.conf.AlluxioConfiguration;
+import alluxio.conf.PropertyKey;
+import alluxio.network.protocol.databuffer.NioDirectBufferPool;
+import alluxio.resource.CloseableResource;
+import alluxio.underfs.UfsManager;
+import alluxio.underfs.UnderFileSystem;
+import alluxio.underfs.options.OpenOptions;
+import alluxio.util.IdUtils;
+import alluxio.worker.block.UfsInputStreamCache;
+import alluxio.worker.block.io.BlockReader;
+import alluxio.worker.block.meta.BlockMeta;
+
+import com.google.common.base.Preconditions;
+import io.netty.buffer.ByteBuf;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.ByteBuffer;
+import java.nio.channels.Channels;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.ReadableByteChannel;
+
+/**
+ * Block reader that reads from UFS.
+ */
+public class PagedUfsBlockReader extends BlockReader {
+  private static final ByteBuffer EMPTY_BYTE_BUFFER = ByteBuffer.allocate(0);
+  private final long mPageSize;
+  private final UfsManager mUfsManager;
+  private final UfsInputStreamCache mUfsInStreamCache;
+  private final BlockMeta mBlockMeta;
+  private final UfsBlockReadOptions mUfsBlockOptions;
+  private final long mInitialOffset;
+  private final ByteBuffer mLastPage;
+  private long mLastPageIndex = -1;
+  private boolean mClosed = false;
+  private long mPosition;
+
+  /**
+   * @param ufsManager
+   * @param ufsInStreamCache
+   * @param conf
+   * @param blockMeta
+   * @param offset
+   * @param ufsBlockReadOptions
+   */
+  public PagedUfsBlockReader(UfsManager ufsManager,
+      UfsInputStreamCache ufsInStreamCache, AlluxioConfiguration conf, BlockMeta blockMeta,
+      long offset, UfsBlockReadOptions ufsBlockReadOptions) {
+    Preconditions.checkArgument(offset >= 0 && offset <= blockMeta.getBlockSize(),
+        "Attempt to read block %s which is %s bytes long at invalid byte offset %s",
+        blockMeta.getBlockId(), blockMeta.getBlockSize(), offset);
+    mUfsManager = ufsManager;
+    mUfsInStreamCache = ufsInStreamCache;
+    mBlockMeta = blockMeta;
+    mUfsBlockOptions = ufsBlockReadOptions;
+    mPageSize = conf.getBytes(PropertyKey.USER_CLIENT_CACHE_PAGE_SIZE);
+    mInitialOffset = offset;
+    mLastPage = ByteBuffer.allocateDirect((int) mPageSize);
```

Comment:
as discussed offline, this is an optimization to avoid frequently open/close file. We would like to further improve this in the future by changing the block reader api.